homeassistant/components/schedule/__init__.py

```diff
@@ -0,0 +1,310 @@
+        self._config = STORAGE_SCHEMA(config)
+        self._attr_capability_attributes = {ATTR_EDITABLE: editable}
+        self._attr_icon = self._config.get(CONF_ICON)
+        self._attr_name = self._config[CONF_NAME]
+        self._attr_unique_id = self._config[CONF_ID]
+
+    @classmethod
+    def from_yaml(cls, config: ConfigType) -> Schedule:
+        """Return entity instance initialized from yaml storage."""
+        schedule = cls(config, editable=False)
+        schedule.entity_id = f"{DOMAIN}.{config[CONF_ID]}"
+        return schedule
+
+    async def async_update_config(self, config: ConfigType) -> None:
+        """Handle when the config is updated."""
+        self._config = STORAGE_SCHEMA(config)
+        self._attr_icon = config.get(CONF_ICON)
+        self._attr_name = config[CONF_NAME]
+        self._clean_up_listener()
+        self._update()
+
+    @callback
+    def _clean_up_listener(self) -> None:
+        """Remove the update timer."""
+        if self._unsub_update is not None:
+            self._unsub_update()
+            self._unsub_update = None
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added to hass."""
+        self.async_on_remove(self._clean_up_listener)
+        self._update()
+
+    @callback
+    def _update(self, _: datetime | None = None) -> None:
+        """Update the states of the schedule."""
+        now = dt_util.now()
+        todays_schedule = self._config.get(WEEKDAY_TO_CONF[now.weekday()], [])
+
+        # Determine current schedule state
+        self._attr_state = next(
+            (
+                STATE_ACTIVE
+                for time_range in todays_schedule
+                if time_range[CONF_FROM] <= now.time() <= time_range[CONF_TO]
+            ),
+            STATE_INACTIVE,
+        )
+
+        # Find next event in the schedule
+        next_event = None
+        for day in range(7):
+            times = sorted(
+                itertools.chain(
+                    *[
+                        [time_range[CONF_FROM], time_range[CONF_TO]]
+                        for time_range in self._config.get(
+                            WEEKDAY_TO_CONF[(now.weekday() + day) % 7], []
+                        )
+                    ]
+                )
+            )
+            if next_event := next(
+                (
+                    possible_next_event
+                    for time in times
+                    if (
+                        possible_next_event := (
+                            datetime.combine(now.date(), time, tzinfo=now.tzinfo)
+                            + timedelta(days=day)
+                        )
+                    )
+                    > now
+                ),
+                None,
+            ):
+                break
+
+        self._attr_extra_state_attributes = {
+            ATTR_NEXT_EVENT: next_event,
+        }
+        self.async_write_ha_state()
+
+        if next_event:
```

Comment:
I was going to comment that midnight is evaluated as `False` but looks like that was [fixed in Python 3.8](https://bugs.python.org/issue13936) 👍